### DISCUSSION TITLE
When was support for existing PDO instance removed?

### DISCUSSION THREAD

**Discussion by duzenko** [doctrine/doctrine2#10835] [Support Questions] [answered]
This issue says it was removed in 3.0 https://github.com/doctrine/dbal/issues/4504
However I'm trying to upgrade to 2.10 and it's giving be a strange error "The options 'driver' or 'driverClass' are mandatory if no PDO instance is given to DriverManager::getConnection()." - I can't see where I'm supposed to pass the instance though. The code that works with 2.9 is this:
```
			$conn = array(
				'pdo' => $GLOBALS[ 'Edb' ]->getPdo()
                        );
...
		$eventManager = new \Doctrine\Common\EventManager();
		$eventManager->addEventListener( $events, $listener );

		$config = Setup::createAnnotationMetadataConfiguration( $entityPaths, $isDevMode, null, null, false );
		$config->addCustomDatetimeFunction( 'MONTH', Month::class );
		$config->addCustomDatetimeFunction( 'YEAR', Year::class );
		$config->addCustomDatetimeFunction( 'ROUND', Round::class );
		$config->addCustomDatetimeFunction( 'IF', IfElse::class );
		$config->setProxyDir( $this->getApplicationPath() . 'DoctrineProxies' );
		$this->entityManager = EntityManager::create( $conn, $config, $eventManager );
```

**Comment by derrabus** [comment]
To answer your question: 3.0 is correct. The feature still exists in the final 2.13 version of DBAL 2.

Apart from that, you're on your own as we don't maintain or support DBAL 2 anymore.

**Reply by duzenko** [reply]
@derrabus so what changed in 2.10 so that the code above stopped working and how are we supposed to pass PDO instances since 2.10?

**Reply by derrabus** [reply]
All I can tell you is that this feature has not been removed or deliberately broken in 2.10. I would recommend to jump to 2.13.9 directly. Maybe that solves the issue already. Maybe not. 🤷‍♂️

**Reply by duzenko** [reply]
@derrabus AFAICS 2.13 throws the same error
Where do I see example in the docs on how to correctly use the existing PDO instance?

**Reply by derrabus** [reply]
This is the documentation that we have: https://www.doctrine-project.org/projects/doctrine-dbal/en/2.13/reference/configuration.html

If that does not help, I'm afraid you'll need to read the source code. Maybe, you can also identify the commit that broke your use-case with `git bisect` and apply a fix to your own fork of DBAL.

**Reply by duzenko** [reply]
@derrabus You probably think that you're being helpful but in fact you aren't

**Reply by derrabus** [reply]
I'm not trying to solve your problem. I'm telling you that you're running an ancient version that we don't maintain anymore and give you the pointers you need to maintain your own fork for your project. Because that's what you need to do to get stuff fixed, whether you like it or not. If that's not enough for you, I'll stop wasting my freetime on you and just close your issues.

**Comment by duzenko** [comment, marked as answer]
In case more people run into this: when upgrading to 2.10 you need to manually restrict another package version:
```
        //"doctrine/orm": "2.8.2",
        "doctrine/orm": "2.13.*",
        "doctrine/dbal": "2.*",
        "doctrine/annotations": "*",
        "symfony/cache": "*",
```
